src/crimson/osd/ops_executer.cc

```diff
@@ -89,29 +89,44 @@ OpsExecuter::call_errorator::future<> OpsExecuter::do_op_call(OSDOp& osd_op)
     [this, prev_rd, prev_wr, &osd_op, flags]
     (auto outcome) -> call_errorator::future<> {
       auto& [ret, outdata] = outcome;
+      osd_op.rval = ret;
+
       logger().debug("do_op_call: method returned ret={}, outdata.length()={}"
                      " while num_read={}, num_write={}",
                      ret, outdata.length(), num_read, num_write);
       if (num_read > prev_rd && !(flags & CLS_METHOD_RD)) {
         logger().error("method tried to read object but is not marked RD");
+        osd_op.rval = -EIO;
         return crimson::ct_error::input_output_error::make();
       }
       if (num_write > prev_wr && !(flags & CLS_METHOD_WR)) {
         logger().error("method tried to update object but is not marked WR");
+        osd_op.rval = -EIO;
         return crimson::ct_error::input_output_error::make();
       }
-
-      // for write calls we never return data expect errors. For details refer
-      // to cls/cls_hello.cc.
-      if (ret < 0 || (flags & CLS_METHOD_WR) == 0) {
-        logger().debug("method called response length={}", outdata.length());
+      // for write calls we never return data expect errors or RETURNVEC.
+      // olease refer cls/cls_hello.cc to details.
+      // ceph-osd has this implemented in `PrimaryLogPG::execute_ctx`,
+      // grep for `ignore_out_data`.
+      if (!op_info->may_write() || op_info->allows_returnvec() || ret < 0) {
```

Comment:
i don't think this condition is correct. IIUC, the reason why  we want to limit the size of  `outdata` is that we will need to persist the list of <retval, outdata> pair to disk as part of pg_log_entry. and pg_log is supposed to be light weight. so any osd op which claims to use `RETURNVEC` but actually will lead to relatively large returned outdata violates the agreement, and will get a `-EOVERFLOW` instead. and the reason why we want to persist these informations using pg_log is that we think this op will have side effects. that's why we have following check in classic osd,

```c++
  if (!ctx->op_t->empty() &&
      op->may_write() &&
      result >= 0) {
    // successful update
    if (ctx->op->allows_returnvec()) {
      // enforce reasonable bound on the return buffer sizes
```

where we perform extra checks only if

- the op has side effect (op_t, i.e., the transaction is not empty), and
- the op may writes, and
- the operation succeeds

i don't quite understand why we need to check if the transaction is empty **and** if the op may write,  but i think the idea is to check the outdata only if we are going to write the pglog because the change caused by the op is committed.

so in this change we should put something like
```c++
if (op_info->may_write() && op_info->allows_returnvec() && ret >= 0) {
```